### PR TITLE
Add support for rsa pss

### PR DESCRIPTION
**Pull request by joYyHack**

### CODE REVIEW

File: contracts/utils/RSAPSS.sol [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        if (s_.length == 0 || e_.length == 0 || n_.length == 0) {
+            return false;
+        }
+
+        bytes memory decipher_ = decrypt(s_, e_, n_);
+
+        return pss(message_, decipher_);
+    }
+
+    function decrypt(
```
Comment (Arvolear) [inline contracts/utils/RSAPSS.sol:28]: Why not use existing `RSA` lib for that?

File: contracts/utils/RSAPSS.sol [hunk 2]
```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.16;
+
+error MaskTooLengthy();
+
+library RSAPSS {
+    uint256 constant HASH_LEN = 32;
+    uint256 constant SALT_LEN = 32;
+    uint256 constant MS_BITS = 4095 & 0x7;
```
Comment (Arvolear) [inline contracts/utils/RSAPSS.sol:9]:
```suggestion
    uint256 constant MS_BITS = 4095;
```

File: contracts/utils/RSAPSS.sol (hunk 1 above)
Comment (joYyHack) [reply contracts/utils/RSAPSS.sol]: moved to rsa, in this case should we use the proper function naming?
As "decrypt" and "verify" functions don't give me the real understanding which function is supposed to be for RSA and RSAPSS

File: contracts/utils/RSAPSS.sol [hunk 3]
```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.16;
+
+import {RSA} from "./RSA.sol";
+
+error MaskTooLengthy();
```
Comment (Arvolear) [inline contracts/utils/RSAPSS.sol:6]:
```suggestion
```

File: contracts/utils/RSAPSS.sol [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    function mgf(
+        bytes memory message_,
+        uint256 maskLen_
+    ) private pure returns (bytes memory res_) {
+        bytes memory cnt_ = new bytes(4);
+
+        if (maskLen_ > (2 ** 32) * HASH_LEN) {
+            revert MaskTooLengthy();
+        }
```
Comment (Arvolear) [inline contracts/utils/RSAPSS.sol:107]:
```suggestion
        require(maskLen_ <= (2 ** 32) * HASH_LEN, "RSAPSS: mask too lengthy");
```

File: deploy/1_state.migration.ts [hunk 5]
```diff
@@ -1,45 +1,26 @@
 import { Deployer, Reporter } from "@solarity/hardhat-migrate";
-import { deployPoseidons } from "./helpers/helper";
+import { deployPoseidons, deployProxy, deploySMTProxy } from "./helpers/helper";
 
-import { StateKeeperMock__factory, PoseidonSMT__factory, ERC1967Proxy__factory } from "@ethers-v6";
+import { PoseidonSMT, StateKeeperMock, StateKeeperMock__factory } from "@ethers-v6";
 
 import { getConfig } from "./config/config";
 
+const SMTInit = async (smt: PoseidonSMT, stateKeeper: StateKeeperMock, config: any) => {
```
Comment (Arvolear) [inline deploy/1_state.migration.ts:8]:
```suggestion
const smtInit = async (smt: PoseidonSMT, stateKeeper: StateKeeperMock, config: any) => {
```

File: deploy/2_registration.migration.ts [hunk 6]
```diff
@@ -1,29 +1,55 @@
...
-  PInternalVerifier2__factory,
+  Registration2Mock__factory,
+  StateKeeperMock__factory,
 } from "@ethers-v6";
 
+import { BaseContract } from "ethers";
 import { getConfig } from "./config/config";
+import { deployProxy } from "./helpers/helper";
+
+const deployDispatcher = async (
```
Comment (Arvolear) [inline deploy/2_registration.migration.ts:23]:
```suggestion
const deployCDispatcher = async (
```

**Review by Arvolear — APPROVED** [review]
That will work

**Pull request merged**